### ISSUE TITLE
Route Objects/Logging/Metrics to Index based on Namespace

### ISSUE THREAD

**Issue by SuitableProduce** [splunk/splunk-connect-for-kubernetes#530]
I'm looking to do two things:
Part A) Looking to route Object, Metrics, and Logging data to specific indexes based on the namespace they come from.
Part B) Additionally, I'd like to be able to modify the index that it is routed to based on the namespace the k8s object originates from.  So, if namespaceA, route to index ABC123.

I've created all Indexes and associated them with the token I've been using. 

I see in issue #376 someone mentioned using annotations at the pod or namespace level.  However, from what I can tell you loose the information gleened from the sourcetype.  Also, if that were the solution, how would you modify the fluentd Record to satisfy part B. Would that also only satisfy the routing of logs and metrics?

Issue #59 seems like it is applicable, but I must be missing a few pieces because I've tried implementing what he posted with no success. If a values.yaml were to be attached with his final resolution, that might clear some things up.  He also states that he is routing logging data using | .namespace = $parts[1].  I'm not sure how to implement that for logs.  Would this also work for metrics?

In general, I'm a bit confused because it seems like the config options stated in the values.yaml/README are missing something. 
 There are a few issues raised here regarding routing data to Index based on namespace, and people are providing snippets of fluentd configs on how they accomplished it - stuff that is not mentioned in the values.yaml/README.  However, at the same time there is a indexRouting option for objects and logs that you can enable.  It states you need to configure customFilters, and provides one example.  Issue #224 is an example of this, but the issue being resolved via an email with no follow up on the thread. So, if the indexRouting option works, why are there other people writting customFilters to route objects/logs to an index based on Namespace? The README states you can use annotations to route logs to an specific index.  But what about objects and metics?

To summarize:
1.  Would it be possible to provide an example config of using indexRouting (or indexRouting + neccesary fluentd customFilters) for Objects, Metrics, and Logging? 
2. It seems like more configuration changes are required in addition to what is mentioned in the README/comments section of the values.yaml.  Could the README be updated with some more clarification on how to accomplish this?

**Comment by SuitableProduce** [comment]
While I do think this is a documentation issue, I'm also actually curious if someone might be able to provide a sample config here.  I've actually been trying to figure this out for a while and any insight would be greatly appreciated.

**Comment by chaitanyaphalak** [comment]
hey @SuitableProduce which version of SCK are you working with. I would recommend you to upgrade to the latest version of SCK - https://github.com/splunk/splunk-connect-for-kubernetes/releases/tag/1.4.4
On this latest version for using indexrouting for logging you can use annotations - https://github.com/splunk/splunk-connect-for-kubernetes#managing-sck-log-ingestion-by-using-annotations
For metrics and objects indexrouting you have to specify custom filters (customFilters). Refer https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-metrics/values.yaml#L212 for metrics and https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-objects/values.yaml#L249 for objects.

In short:
For index routing for logs use annotations with https://github.com/splunk/splunk-connect-for-kubernetes#managing-sck-log-ingestion-by-using-annotations
For index routing for metrics use https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-metrics/values.yaml#L212
For index routing for metrics use https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-objects/values.yaml#L249

**Comment by BrandonTsai** [comment]
@chaitanyaphalak I have tried the latest version 1.4.4, but I still can not figure out how can I configure "indexRouting" for Objects and Metrics.
I want to forward Objects and Metrics to different indexes for different namespaces.
The links you gave only forward all data to another index, it did not route to different indexes as Logs.

**Comment by SuitableProduce** [comment]
@chaitanyaphalak 
Thanks for the reply and sorry for the delay!  I've been working with the customFilters, but the examples provided in the readme don't account for dynamically sending data to an index based on the namespace the fluentd record originates from.  I can set the index manually in the values.yaml as defined in the comments, but I don't think that will work if I want to route based on record namespace.

My current throught process is to figure out how to set the index using the jq_transformer (or record_transformer, or record_modifier), using the namespace env variable from the fluentd record.

I've tried many variations trying to get this working, two of which are below:

This doesn't work, but I feel ilke its close?
`  NamespaceSourcetypeFilter:
     tag: **
     type: record_transformer
     body: jq '.record.index= "${record["namespace']}"'
`
This sets the index variable to the namespace where the SCK pod is running, not where the record originates from.  So again, close ( think?), but no cigar.

` NamespaceSourcetypeFilter:
    tag **
    type: jq_transformer
    body: jq '.record.index = "\(env.MY_NAMESPACE)" | .record' 
`

I've tried many other variations of those two, all the while trying to figure out the env variable of the origination record.  Identifying that source record namespace variable, and using it to define the index via a jq_transformer, record_transformer, record_modifier, or some other fluentd plugin, is really the crux of my issue.
